### PR TITLE
Memory adaptive flow control

### PR DESCRIPTION
**Pull request by AlexStocks**
### Describe what this PR does / why we need it

adaptive memory slot

### CODE REVIEW

File: core/system/rule_manager.go [hunk 1]
```diff
@@ -26,7 +26,7 @@ type RuleMap map[MetricType][]*Rule
 
 // const
 var (
-	ruleMap    = make(RuleMap)
+	ruleMap    = make(RuleMap, 8)
```
Comment (louyuting) [inline core/system/rule_manager.go:29]: initial size 8 is meaningless. 
refer to https://github.com/golang/go/blob/41d8e61a6b9d8f9db912626eb2bbc535e929fefc/src/runtime/map.go#L289
bucketCnt == 8.

File: core/system/rule.go [hunk 2]
```diff
@@ -71,18 +74,20 @@ func (t AdaptiveStrategy) String() string {
...
 type Rule struct {
-	ID           string           `json:"id,omitempty"`
-	MetricType   MetricType       `json:"metricType"`
-	TriggerCount float64          `json:"triggerCount"`
-	Strategy     AdaptiveStrategy `json:"strategy"`
+	ID            string           `json:"id,omitempty"`
+	MetricType    MetricType       `json:"metricType"`
+	TriggerCount  float64          `json:"triggerCount,omitempty"`
+	Strategy      AdaptiveStrategy `json:"strategy"`
+	LowWaterMark  int64            `json:"lowWaterMark,omitempty"`
```
Comment (louyuting) [inline core/system/rule.go:81]: Maybe we need to keep naming convention about triggered threshold.
such as: "LowThreshold" and "HighThreshold".

File: core/system/slot.go [hunk 3]
```diff
@@ -104,6 +104,15 @@ func (s *AdaptiveSlot) doCheckRule(rule *Rule) (bool, string, float64) {
 			}
 		}
 		return true, "", c
+	case MemUsage:
+		m := CurrentMemoryUsage()
+		if m > rule.LowWaterMark {
+			if rule.Strategy != BBR || !checkBbrSimple() {
```
Comment (louyuting) [inline core/system/slot.go:110]: The implementation of func "checkBbrSimple()" has nothing to do with memory usage.

File: core/flow/rule.go [hunk 4]
```diff
@@ -110,13 +146,43 @@ func (r *Rule) isEqualsTo(newRule *Rule) bool {
...
+	if r.AdaptiveEnable {
+		if r.AdaptiveType == Memory {
+			m := uint64(system.CurrentMemoryUsage())
+			if m <= r.LowWaterMark {
+				threshold = float64(r.SafeThreshold)
+			} else if m >= r.HighWaterMark {
+				threshold = float64(r.RiskThreshold)
+			} else {
+				threshold = float64(((m-r.LowWaterMark)/(r.HighWaterMark-r.LowWaterMark))*
+					(r.RiskThreshold-r.SafeThreshold) + r.SafeThreshold)
```
Comment (louyuting) [inline core/flow/rule.go:178]: r.RiskThreshold-r.SafeThreshold is negative.
it cause overflow for uint64.

File: core/flow/rule.go [hunk 5]
```diff
@@ -110,13 +146,43 @@ func (r *Rule) isEqualsTo(newRule *Rule) bool {
...
+		r.SafeThreshold == newRule.SafeThreshold && r.RiskThreshold == newRule.RiskThreshold &&
+		r.LowWaterMark == newRule.LowWaterMark && r.HighWaterMark == newRule.HighWaterMark) {
+
 		return false
 	}
+
 	return true
 }
 
+func (r *Rule) GetThreshold() float64 {
```
Comment (louyuting) [inline core/flow/rule.go:166]: Maybe we need unit test for this function

File: core/flow/rule.go [hunk 6]
```diff
@@ -102,6 +130,14 @@ type Rule struct {
 	// If the StatIntervalInMs user specifies can not reuse the global statistic of resource,
 	// 		sentinel will generate independent statistic structure for this rule.
 	StatIntervalInMs uint32 `json:"statIntervalInMs"`
+
+	// adaptive flow control alg
+	AdaptiveEnable bool         `json:"adaptiveEnable"`
+	AdaptiveType   AdaptiveType `json:"adaptiveType"`
+	SafeThreshold  uint64       `json:"safeThreshold"`
```
Comment (louyuting) [inline core/flow/rule.go:137]: Maybe we need more detailed commens.

File: core/system/sys_stat.go [hunk 7]
```diff
@@ -86,6 +91,27 @@ func retrieveAndUpdateSystemStat() {
...
 	}
+
+	memoryUsedBytes, err := GetProcessMemoryStat()
+	if err == nil {
+		currentMemoryUsage.Store(memoryUsedBytes)
+	}
+}
+
+var (
+	CurrentPID = os.Getpid()
```
Comment (sanxun0325) [inline core/system/sys_stat.go:102]: CurrentPID可以移到上面统一的var 块中

File: core/system/sys_stat.go [hunk 8]
```diff
@@ -86,6 +91,27 @@ func retrieveAndUpdateSystemStat() {
...
+// GetProcessMemoryStat gets current process's memory usage in Bytes
+func GetProcessMemoryStat() (int64, error) {
+	p, err := process.NewProcess(int32(CurrentPID))
+	if err != nil {
+		return 0, err
+	}
+
+	memInfo, err := p.MemoryInfo()
+
+	return int64(memInfo.RSS), err
```
Comment (sanxun0325) [inline core/system/sys_stat.go:114]: memInfo 有可能为nil,memInfo.RSS将出现错误。

File: core/flow/rule_manager.go [hunk 9]
```diff
@@ -499,5 +501,28 @@ func IsValidRule(rule *Rule) error {
...
+
+		if rule.LowWaterMark == 0 {
+			return errors.New("rule.LowWaterMark should not be 0")
+		}
+
+		if rule.HighWaterMark == 0 {
+			return errors.New("rule.HighWaterMark should not be 0")
+		}
+
+		if rule.LowWaterMark >= rule.HighWaterMark {
```
[the remark this answers is not in the document]
Comment (AlexStocks) [reply core/flow/rule_manager.go:521]: 不行，adaptiveType == Memory 时这个参数值是内存的使用量，所以无法让系统设置。

**Comment by louyuting** [comment]
@AlexStocks  Could you please fix the conflicts?

File: core/flow/rule_manager.go [hunk 10]
```diff
@@ -499,5 +501,28 @@ func IsValidRule(rule *Rule) error {
...
+
+		if rule.RiskThreshold >= rule.SafeThreshold {
+			return errors.New("rule.RiskThreshold should be less than rule.SafeThreshold")
+		}
+
+		if rule.LowWaterMark == 0 {
+			return errors.New("rule.LowWaterMark should not be 0")
+		}
+
+		if rule.HighWaterMark == 0 {
```
[the remark this answers is not in the document]
Comment (AlexStocks) [reply core/flow/rule_manager.go:517]: 有道理

File: core/system/sys_stat.go (hunk 8 above)
Comment (AlexStocks) [reply core/system/sys_stat.go:115]: good advice

File: core/flow/rule_manager.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexStocks) [reply core/flow/rule_manager.go:517]: u r right.

File: core/system/sys_stat_test.go [hunk 11]
```diff
@@ -18,7 +18,7 @@ import (
 	"testing"
 
 	"github.com/alibaba/sentinel-golang/util"
-	"github.com/shirou/gopsutil/v3/cpu"
+	"github.com/shirou/gopsutil/cpu"
```
Comment (sanxun0325) [inline core/system/sys_stat_test.go:21]: Should use github.com/shirou/gopsutil/v3/cpu,shouldn't change it

**Comment by louyuting** [comment]
@AlexStocks  Could you please fix the conflicts?

File: core/flow/rule.go [hunk 12]
```diff
@@ -78,6 +81,33 @@ func (s ControlBehavior) String() string {
 	}
 }
 
+type AdaptiveType uint32
```
Comment (louyuting) [inline core/flow/rule.go:84]: Remove these code block

File: core/flow/rule_manager.go [hunk 13]
```diff
@@ -32,6 +33,8 @@ import (
 // TrafficControllerGenFunc represents the TrafficShapingController generator function of a specific control behavior.
 type TrafficControllerGenFunc func(*Rule, *standaloneStatistic) (*TrafficShapingController, error)
 
+type GetThresholdFunc func() float64
```
Comment (louyuting) [inline core/flow/rule_manager.go:36]: Remove this line

File: core/flow/rule_test.go [hunk 14]
```diff
@@ -20,6 +20,14 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+func TestMericType(t *testing.T) {
```
Comment (louyuting) [inline core/flow/rule_test.go:23]: Remove this ut

File: api/api_test.go [hunk 15]
```diff
@@ -173,3 +179,115 @@ func Test_entryWithArgsAndChainBlock(t *testing.T) {
 	ssm.AssertNumberOfCalls(t, "OnEntryBlocked", 1)
 	ssm.AssertNumberOfCalls(t, "OnCompleted", 0)
 }
+
+func TestAdaptiveFlowControl(t *testing.T) {
```
Comment (louyuting) [inline api/api_test.go:183]: Maybe we need move this test(integration test) to `tests/core`

File: core/config/entity.go [hunk 16]
```diff
@@ -82,7 +82,9 @@ type StatConfig struct {
 // SystemStatConfig represents the configuration items of system statistics.
 type SystemStatConfig struct {
 	// CollectIntervalMs represents the collecting interval of the system metrics collector.
-	CollectIntervalMs uint32 `yaml:"collectIntervalMs"`
+	CollectIntervalMs       uint32 `yaml:"collectIntervalMs"`
```
Comment (louyuting) [inline core/config/entity.go:85]: need CollectLoadIntervalMs field.

File: core/flow/rule.go [hunk 17]
```diff
@@ -102,6 +132,17 @@ type Rule struct {
...
 	// 		sentinel will generate independent statistic structure for this rule.
 	StatIntervalInMs uint32 `json:"statIntervalInMs"`
+
+	// adaptive flow control alg
+	// limitation: SafeThreshold > RiskThreshold && HighWaterMark > LowWaterMark
+	// If the watermark is less than Rule.LowWaterMark, the threshold is Rule.SafeThreshold.
+	// If the watermark is greater than Rule.HighWaterMark, the threshold is Rule.RiskThreshold.
+	// Otherwise, the threshold is ((watermark - LowWaterMark)/(HighWaterMark - LowWaterMark)) *
+	//	(SafeThreshold - RiskThreshold) + RiskThreshold.
+	SafeThreshold int64 `json:"safeThreshold"`
```
Comment (louyuting) [inline core/flow/rule.go:142]: Currently, these fields are related to memory adaptation, but the field names do not give the related meaning. I don't have a better suggestion at the moment. We may need more discussion.

File: core/flow/tc_adaptive.go [hunk 18]
```diff
@@ -0,0 +1,39 @@
+package flow
+
+import "github.com/alibaba/sentinel-golang/core/system_metric"
+
+type MemoryAdaptiveTrafficShapingCalculator struct {
```
Comment (louyuting) [inline core/flow/tc_adaptive.go:5]: need more comments for this TrafficShaping token Calculator

File: metrics/metrics.go [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+		return
+	}
+	registerMetrics.Do(func() {
+		for _, metric := range metrics {
+			registry.MustRegister(metric)
+		}
+	})
+}
+
+type resettable interface {
```
Comment (louyuting) [inline metrics/metrics.go:109]: Could prometheus guarantee all metric implement this interface?

File: metrics/metrics.go [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+	Reset()
+}
+
+// Reset all metrics to zero
+func ResetSentinelMetrics() {
+	for _, metric := range metrics {
+		rm := metric.(resettable)
```
Comment (louyuting) [inline metrics/metrics.go:116]: Maybe we need check ok?

File: core/flow/rule_manager.go [hunk 21]
```diff
@@ -513,5 +546,28 @@ func IsValidRule(rule *Rule) error {
 	if rule.StatIntervalInMs > 10*60*1000 {
 		logging.Info("StatIntervalInMs is great than 10 minutes, less than 10 minutes is recommended.")
 	}
+	if rule.TokenCalculateStrategy == AdaptiveMemory {
+		if rule.RiskThreshold >= rule.SafeThreshold {
+			return errors.New("rule.RiskThreshold should be less than rule.SafeThreshold")
+		}
+
+		if rule.LowWaterMark == 0 {
```
Comment (liqiangz) [inline core/flow/rule_manager.go:554]: Maybe we should check HighWaterMark , LowWaterMark, RiskThreshold, SafeThreshold is not negative.

File: core/flow/rule.go [hunk 22]
```diff
@@ -110,10 +120,19 @@ func (r *Rule) isEqualsTo(newRule *Rule) bool {
...
-		r.MaxQueueingTimeMs == newRule.MaxQueueingTimeMs && r.WarmUpPeriodSec == newRule.WarmUpPeriodSec && r.WarmUpColdFactor == newRule.WarmUpColdFactor) {
+
+		r.TokenCalculateStrategy == newRule.TokenCalculateStrategy && r.ControlBehavior == newRule.ControlBehavior &&
+		util.Float64Equals(r.Threshold, newRule.Threshold) &&
+
+		r.MaxQueueingTimeMs == newRule.MaxQueueingTimeMs && r.WarmUpPeriodSec == newRule.WarmUpPeriodSec &&
+		r.WarmUpColdFactor == newRule.WarmUpColdFactor &&
+
+		r.SafeThreshold == newRule.SafeThreshold && r.RiskThreshold == newRule.RiskThreshold &&
+		r.LowWaterMark == newRule.LowWaterMark && r.HighWaterMark == newRule.HighWaterMark) {
```
Comment (liqiangz) [inline core/flow/rule.go:131]: Direct token calculate strategy and throttling control behavior don't use stat,  needStatistic() function should be modified.

File: core/flow/rule_manager.go (hunk 21 above)
Comment (louyuting) [reply core/flow/rule_manager.go:554]: updated

File: core/flow/rule.go (hunk 22 above)
Comment (louyuting) [reply core/flow/rule.go:131]: updated

File: core/flow/rule.go [hunk 23]
```diff
@@ -80,6 +83,7 @@ func (s ControlBehavior) String() string {
 
 // Rule describes the strategy of flow control, the flow control strategy is based on QPS statistic metric
 type Rule struct {
+	DebugMode bool
```
Comment (louyuting) [inline core/flow/rule.go:86]: Maybe we need remove here.

**Review by louyuting — APPROVED** [review]
LGTM

File: core/system_metric/sys_metric_stat.go [hunk 24]
```diff
@@ -15,50 +15,79 @@
...
+	currentLoad.Store(NotRetrievedLoadValue)
+	currentCpuUsage.Store(NotRetrievedCpuUsageValue)
+	currentMemoryUsage.Store(NotRetrievedMemoryValue)
+
+	p, err := process.NewProcess(int32(CurrentPID))
+	if err != nil {
+		logging.Error(err, "Fail to new process when initializing system metric", "pid", CurrentPID)
+	}
+	currentProcessOnce.Do(func() {
+		currentProcess.Store(p)
```
Comment (liqiangz) [inline core/system_metric/sys_metric_stat.go:64]: Maybe we should not set currentProcess when err != nil

Comment (louyuting) [reply core/system_metric/sys_metric_stat.go:64]: Exactly, I will update here

File: core/system_metric/sys_metric_stat.go [hunk 25]
```diff
@@ -146,3 +235,17 @@ func CurrentCpuUsage() float64 {
 func SetSystemCpuUsage(cpuUsage float64) {
 	currentCpuUsage.Store(cpuUsage)
 }
+
+func CurrentMemoryUsage() int64 {
+	bytes, ok := currentMemoryUsage.Load().(int64)
+	if !ok {
+		return NotRetrievedMemoryValue
```
Comment (liqiangz) [inline core/system_metric/sys_metric_stat.go:242]: Maybe we should check if we can retrieve when we loadRules

File: core/flow/tc_adaptive.go [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+	}
+}
+
+func (m *MemoryAdaptiveTrafficShapingCalculator) BoundOwner() *TrafficShapingController {
+	return m.owner
+}
+
+func (m *MemoryAdaptiveTrafficShapingCalculator) CalculateAllowedTokens(_ uint32, _ int32) float64 {
+	var threshold float64
+	mem := system_metric.CurrentMemoryUsage()
```
Comment (liqiangz) [inline core/flow/tc_adaptive.go:36]: We should check if we get CurrentMemoryUsage.

File: core/system_metric/sys_metric_stat.go [hunk 27]
```diff
@@ -15,50 +15,79 @@
...
+		currentProcess.Store(p)
+	})
 }
 
-func InitCollector(intervalMs uint32) {
+// getMemoryStat returns the current machine's memory statistic
+func getTotalMemorySize() (total uint64) {
+	stat, err := mem.VirtualMemory()
+	if err != nil {
+		return 0
```
Comment (liqiangz) [inline core/system_metric/sys_metric_stat.go:72]: Maybe we can log err When we get VirtualMemory.

File: core/system_metric/sys_metric_stat.go [hunk 28]
```diff
@@ -68,63 +97,123 @@ func InitCollector(intervalMs uint32) {
...
-			prevUsed := calculateUserCpuTick(prev) + calculateKernelCpuTick(prev)
-			curUsed := calculateUserCpuTick(curCpuStat) + calculateKernelCpuTick(curCpuStat)
-			cpuUsage = (curUsed - prevUsed) / tDiff
-			cpuUsage = math.Max(0.0, cpuUsage)
-			cpuUsage = math.Min(1.0, cpuUsage)
-		}
-		currentCpuUsage.Store(cpuUsage)
+func retrieveAndUpdateCpuStat() {
+	cpuPercent, err := getProcessCpuStat()
+	if err == nil {
```
Comment (liqiangz) [inline core/system_metric/sys_metric_stat.go:156]: Maybe we can log err here.

File: core/system_metric/sys_metric_stat.go [hunk 29]
```diff
@@ -68,63 +97,123 @@ func InitCollector(intervalMs uint32) {
 	})
 }
 
-func retrieveAndUpdateSystemStat() {
-	cpuStats, err := cpu.Times(false)
-	if err != nil {
-		logging.Warn("[retrieveAndUpdateSystemStat] Failed to retrieve current CPU usage", "err", err.Error())
+func retrieveAndUpdateMemoryStat() {
+	memoryUsedBytes, err := GetProcessMemoryStat()
+	if err == nil {
```
Comment (liqiangz) [inline core/system_metric/sys_metric_stat.go:102]: Maybe we can log err here.

File: core/flow/rule.go [hunk 30]
```diff
@@ -102,6 +105,16 @@ type Rule struct {
...
+
+	// adaptive flow control algorithm related parameters
+	// limitation: SafeThreshold > RiskThreshold && HighWaterMark > LowWaterMark
+	// if the current memory usage is less than or equals to LowWaterMark, threshold == SafeThreshold
+	// if the current memory usage is more than or equals to HighWaterMark, threshold == RiskThreshold
+	// if  the current memory usage is in (LowWaterMark, HighWaterMark), threshold is in (RiskThreshold, SafeThreshold)
+	SafeThreshold int64 `json:"safeThreshold"`
+	RiskThreshold int64 `json:"riskThreshold"`
+	LowWaterMark  int64 `json:"lowWatermark"`
+	HighWaterMark int64 `json:"highWatermark"`
```
Comment (luckyxiaoqiang) [inline core/flow/rule.go:117]: How about below
```go
LowMemUsageThreshold
HighMemUsageThreshold
MemLowWaterMark
MemHighWaterMark
```

File: core/system_metric/sys_metric_stat.go (hunk 25 above)
Comment (louyuting) [reply core/system_metric/sys_metric_stat.go:242]: User need to guarantee open memory collector

File: core/flow/tc_adaptive.go (hunk 26 above)
Comment (louyuting) [reply core/flow/tc_adaptive.go:36]: in system_metric module, we need another PR to enhance.
Could you please file a Issue?

File: core/flow/rule.go (hunk 30 above)
Comment (louyuting) [reply core/flow/rule.go:117]: I think it's ok.

File: core/flow/tc_adaptive.go (hunk 26 above)
Comment (liqiangz) [reply core/flow/tc_adaptive.go:39]: ok

**Review by sanxun0325 — APPROVED** [review]

**Review by louyuting — DISMISSED** [review]
LGTM

File: core/flow/tc_adaptive_test.go [hunk 31]
```diff
@@ -0,0 +1,28 @@
...
+	assert.True(t, int64(tc1.CalculateAllowedTokens(0, 0)) == tc1.lowMemUsageThreshold)
+	system_metric.SetSystemMemoryUsage(1024)
+	assert.True(t, int64(tc1.CalculateAllowedTokens(0, 0)) == tc1.lowMemUsageThreshold)
+	system_metric.SetSystemMemoryUsage(1536)
+	assert.True(t, int64(tc1.CalculateAllowedTokens(0, 0)) == 550)
+	system_metric.SetSystemMemoryUsage(2048)
+	assert.True(t, int64(tc1.CalculateAllowedTokens(0, 0)) == 100)
+	system_metric.SetSystemMemoryUsage(3072)
+	assert.True(t, int64(tc1.CalculateAllowedTokens(0, 0)) == 100)
+}
```
Comment (luckyxiaoqiang) [inline core/flow/tc_adaptive_test.go:28]: ·util.Float64Equals· may be better.

File: core/flow/rule.go [hunk 32]
```diff
@@ -102,6 +105,16 @@ type Rule struct {
...
+
+	// adaptive flow control algorithm related parameters
+	// limitation: LowMemUsageThreshold > HighMemUsageThreshold && MemHighWaterMark > MemLowWaterMark
+	// if the current memory usage is less than or equals to MemLowWaterMark, threshold == LowMemUsageThreshold
+	// if the current memory usage is more than or equals to MemHighWaterMark, threshold == HighMemUsageThreshold
+	// if  the current memory usage is in (MemLowWaterMark, MemHighWaterMark), threshold is in (HighMemUsageThreshold, LowMemUsageThreshold)
+	LowMemUsageThreshold  int64 `json:"lowMemUsageThreshold"`
+	HighMemUsageThreshold int64 `json:"highMemUsageThreshold"`
+	MemLowWaterMark       int64 `json:"memLowWatermark"`
+	MemHighWaterMark      int64 `json:"memHighWatermark"`
```
Comment (luckyxiaoqiang) [inline core/flow/rule.go:117]: Add a unit suffix may be better, like `KB`, `MB`.

File: tests/api/api_entry_integration_test.go [hunk 33]
```diff
@@ -0,0 +1,127 @@
...
+		entry.Exit()
+	}
+	_, blockError := api.Entry(rs, api.WithTrafficType(base.Inbound))
+	assert.NotNil(t, blockError)
+	if blockError != nil {
+		t.Logf("entry error:%+v, caused: %+v", blockError.Error(), blockError.TriggeredRule())
+	}
+
+	// clear statistic
+	time.Sleep(time.Second * 2)
```
Comment (luckyxiaoqiang) [inline tests/api/api_entry_integration_test.go:70]: Use mock time in util pkg may be better.

**Comment by sczyh30** [comment]
@AlexStocks Could you please sign the CLA here: https://cla-assistant.io/alibaba/sentinel-golang?pullRequest=348

File: core/flow/tc_adaptive_test.go (hunk 31 above)
Comment (louyuting) [reply core/flow/tc_adaptive_test.go:28]: updated

File: core/flow/rule.go (hunk 32 above)
Comment (louyuting) [reply core/flow/rule.go:117]: Will use Bytes and provide convenient MB KB constant

**Review by luckyxiaoqiang — APPROVED** [review]
LGTM.

File: core/base/block_error.go [hunk 34]
```diff
@@ -65,6 +65,10 @@ func NewBlockErrorWithCause(blockType BlockType, blockMsg string, rule SentinelR
 }
 
 func (e *BlockError) Error() string {
+	if e == nil {
```
Comment (sczyh30) [inline core/base/block_error.go:68]: Does it make sense here?

File: core/config/entity.go [hunk 35]
```diff
@@ -82,7 +82,10 @@ type StatConfig struct {
...
+	CollectLoadIntervalMs   uint32 `yaml:"collectLoadIntervalMs"`
```
Comment (sczyh30) [inline core/config/entity.go:86]: Could you please add some doc comments in the new fields?

Comment (louyuting) [reply core/config/entity.go:86]: Okay

**Review by louyuting — APPROVED** [review]
LGTM

**Pull request merged**